Return null on allocation failure due to OOM instead of throwing a exception from within the GC

swgillespie opened the pull request:
The idea of throwing an `OutOfMemoryException` on allocation failure is a VM-specific idea that shouldn't be within the GC - the function `ThrowOutOfMemory` is itself defined in `src/vm/gchelpers.cpp` and shouldn't be called from the GC side of the GC-EE interface.

This PR aligns the CoreCLR GC to the Redhawk strategy of returning `null` on allocation failure and letting the VM decide what to do when an allocation fails. In this case, the VM does exactly what the GC is doing today - it throws an out of memory exception if any of the allocation routines return `null`. This has the benefit of removing some `#ifdef FEATURE_REDHAWK`s in the GC, aligning CoreRT and CoreCLR's allocation strategy, and removing the interface violation.

The only difference in behavior that this will cause is that allocating an object that is too large will not call `ThrowOutOfMemoryDimensionsExceeded` like it does today - I am not sure how CoreRT deals with this and what it affects. Currently, that function calls `ThrowOutOfMemory` unless `_WIN64` is defined, otherwise it throws an `OutOfMemoryException` with a specific error message.

cc @jkotas, who suggested this.

jkotas commented:
> too large will not call ThrowOutOfMemoryDimensionsExceeded like it does today

IIRC, this exception was originally introduced to reduce our support cost. We got repeated questions about why is my program throwing out of memory exception while I have a plenty of memory available - in some cases, the actual problem was that the array asked for was bigger than the internal limits. I think we should keep throwing this exception on the VM side as appropriate - move the `GetGCAllowVeryLargeObjects` config switch and related code to the VM side as well.

CoreRT does not distinguish between this case today, but it is more like a bug.

Otherwise looks good.

swgillespie commented:
cc @Maoni0 @sergiy-k @adityamandaleeka

swgillespie commented:
@jkotas done, thanks!

I left most of the code checking for too-large objects in the GC for assertion purposes - I'm happy to do something else if asked.

jkotas commented on src/gc/gc.cpp:
```diff
@@ -30484,6 +30484,9 @@ CObjectHeader* gc_heap::allocate_large_object (size_t jsize, int64_t& alloc_byte
+    // these next few lines are not strictly necessary anymore - they are here
+    // to sanity check that we didn't get asked to create an object
+    // that's too large.
     size_t maxObjectSize = (INT32_MAX - 7 - Align(min_obj_size));
```
I think you should be able to simplify it to:

```
#if BIT64
size_t maxObjectSize = (INT64_MAX - 7 - Align(min_obj_size)); 
#else
size_t maxObjectSize = (INT32_MAX - 7 - Align(min_obj_size)); 
#endif
```

And remove `GetGCAllowVeryLargeObjects`.

swgillespie replied:
ok!

jkotas commented on src/vm/gchelpers.cpp:
```diff
@@ -60,6 +60,40 @@ inline gc_alloc_context* GetThreadAllocContext()
+// the OOM was not from memory pressure but from an object
+// being too large.
+inline void CheckObjectSizeOrThrow(size_t alloc_size)
+{
```
This may need:
```
 CONTRACTL {
        THROWS;
        GC_TRIGGERS;
    } CONTRACTL_END;
```

to prevent the static contract verifier in internal build from complaining.

swgillespie replied:
ok!

jkotas commented on src/vm/gchelpers.cpp:
```diff
@@ -60,6 +60,40 @@ inline gc_alloc_context* GetThreadAllocContext()
+// an OutOfMemoryException with a message indicating that
+// the OOM was not from memory pressure but from an object
+// being too large.
+inline void CheckObjectSizeOrThrow(size_t alloc_size)
```
Nit - naming the method just `CheckObjectSize` may look better.

swgillespie replied:
will do!

Maoni0 commented on src/vm/gchelpers.cpp:
```diff
@@ -60,6 +60,45 @@ inline gc_alloc_context* GetThreadAllocContext()
+        GC_TRIGGERS;
+    } CONTRACTL_END;
+
+    size_t min_obj_size = sizeof(uint8_t*)   // sync block
```
first of all, this calculation is incorrect, min_obj_size is 3-ptr size, not 4; 2ndly, this should not be calculated like this, it should be obtained from the GC side and for the places in the VM that assumes the min_obj_size, we should assert that they agree.

swgillespie replied:
> first of all, this calculation is incorrect, min_obj_size is 3-ptr size, not 4;

You're right, the error here is in the alignment - I misread the `ALIGNCONST` macro, I apologize.

How would you like this to be exposed? `LARGE_OBJECT_SIZE` is already `#defined` in `gcinterface.h`, so would you be okay with defining `min_obj_size` there?

Maoni0 replied:
hmm, I don't see how this is related to ALIGNCONST... 

yes, all the sizes defined by the GC that're used by vm should be exposed via the interface so you can add this to gcinterface.h.

swgillespie replied:
@Maoni0 it's the default alignment parameter to `Align`. Anyway, will do!

jkotas commented on src/gc/gcinterface.h:
```diff
@@ -357,7 +357,17 @@ class IGCHeap {
+    These allocation routines should not be called with allocation requests
+    larger than:
+       32-bit                                 -> 0x7FFFFFE0
+       64-bit, GCAllowVeryLargeObjects: false -> 0x000000007FFFFFE0
```
GCAllowVeryLargeObjects should be omitted here. It is internal limitation imposed by the CLR VM; not part of the GC interface.

swgillespie replied:
ok!

swgillespie commented:
The last commit addresses all CR feedback except Maoni's feedback about allocation alignment throughout the VM - in our assembly helpers we use immediates instead of defined constants to align allocations, so we could replace all of the immediates used in alignment with defined constants and assert that the ones exported by the GC are the same. I'm still working on this feedback.

swgillespie commented:
@jkotas The GC and the the JIT helpers/VM need to agree on alignment requirements and right now they are hardcoded immediates - when introducing constants for them, do you have a place you'd prefer the constants to be checked against each other? The GC would have a ALIGN_CONSTANT that it defines and each `asmconstants.h` would have their own, and somewhere we'd static_assert that they are the same.

jkotas commented:
> do you have a place you'd prefer the constants to be checked against each other

I think that src/vm/gcenv.ee.cpp would be the most appropriate place to assert this consistency. It has checks that verify consistency between the gcinterface constants and the VM constants already, like: https://github.com/dotnet/coreclr/blob/master/src/vm/gcenv.ee.cpp#L30

swgillespie commented:
I spoke to Maoni offline and decided that I'll do the alignment assertion changes in another PR after this one gets merged. I discussed and addressed the review comments offline with Maoni so this is ready to merge.

Maoni0 commented:
:shipit:
